File: src/lib/android-engine/fs.ts
// Android FS engine — uses @tauri-apps/plugin-fs directly.
// No Go binary needed. All file I/O goes through Tauri's native FS plugin.

import {
  readTextFile, writeTextFile, readDir, mkdir, remove,
  rename, copyFile, exists, stat, readFile, writeFile,
} from '@tauri-apps/plugin-fs'

// ── Tree walker ────────────────────────────────────────────────

const SKIP = new Set(['node_modules', '.git', 'dist', 'build', '__pycache__', '.venv', 'venv', '.cache', 'target', 'out', 'vendor'])

export interface FsNode {
  id: string
  name: string
  path: string
  type: 'file' | 'folder'
  children?: FsNode[]
}

async function walkDir(dirPath: string, depth: number, maxDepth: number): Promise<FsNode[]> {
  if (depth > maxDepth) return []
  let entries
  try { entries = await readDir(dirPath) } catch { return [] }
  const nodes: FsNode[] = []
  for (const e of entries) {
    if (!e.name || e.name.startsWith('.') || SKIP.has(e.name)) continue
    const fullPath = `${dirPath}/${e.name}`
    if (e.isDirectory) {
      const children = await walkDir(fullPath, depth + 1, maxDepth)
      nodes.push({ id: fullPath, name: e.name, path: fullPath, type: 'folder', children })
    } else {
      nodes.push({ id: fullPath, name: e.name, path: fullPath, type: 'file' })
    }
  }
  nodes.sort((a, b) => {
    if (a.type !== b.type) return a.type === 'folder' ? -1 : 1
    return a.name.localeCompare(b.name)
  })
  return nodes
}

// ── Flat file lister ───────────────────────────────────────────

async function listAllFlat(dirPath: string, result: string[], maxFiles: number) {
  if (result.length >= maxFiles) return
  let entries
  try { entries = await readDir(dirPath) } catch { return }
  for (const e of entries) {
    if (!e.name || e.name.startsWith('.') || SKIP.has(e.name)) continue
    const fp = `${dirPath}/${e.name}`
    if (e.isDirectory) {
      await listAllFlat(fp, result, maxFiles)
    } else {
      result.push(fp)
      if (result.length >= maxFiles) return
    }
  }
}

// ── Import scanner (lightweight) ───────────────────────────────

function extractImports(code: string, filePath: string): string[] {
  const deps: string[] = []
  const dir = filePath.substring(0, filePath.lastIndexOf('/'))
  for (const m of code.matchAll(/from\s+['"](\.[^'"]+)['"]/g)) deps.push(m[1])
  for (const m of code.matchAll(/import\s+['"](\.[^'"]+)['"]/g)) deps.push(m[1])
  for (const m of code.matchAll(/require\(['"](\.[^'"]+)['"]\)/g)) deps.push(m[1])
  return deps.map(d => {
    const exts = ['.ts', '.tsx', '.js', '.jsx', '.py', '']
    for (const ext of exts) {
      const candidate = `${dir}/${d}${ext}`.replace(/\/\.\//g, '/').replace(/\/[^/]+\/\.\.\//g, '/')
      return candidate
    }
    return `${dir}/${d}`
  })
}

// ── Exported handlers ──────────────────────────────────────────

export async function handleFsTree(rootPath: string, maxDepth = 4) {
  const tree = await walkDir(rootPath, 0, maxDepth)
  return { tree }
}

export async function handleFsRead(filePath: string) {
  try {
    const content = await readTextFile(filePath)
    return { content }
  } catch (e: any) {
    return { error: e?.message ?? 'read failed' }
  }
}

export async function handleFsWrite(filePath: string, content: string) {
  try {
    await writeTextFile(filePath, content)
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleFsCreateFile(filePath: string) {
  try {
    await writeTextFile(filePath, '')
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleFsCreateDir(folderPath: string) {
  try {
    await mkdir(folderPath, { recursive: true })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleFsDelete(itemPath: string) {
  try {
    await remove(itemPath, { recursive: true })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleFsRename(oldPath: string, newPath: string) {
  try {
    await rename(oldPath, newPath)
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleFsCopyFile(srcPath: string, destPath: string) {
  try {
    await copyFile(srcPath, destPath)
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleFsCopyFolder(srcPath: string, destPath: string) {
  try {
    await mkdir(destPath, { recursive: true })
    const entries = await readDir(srcPath)
    for (const e of entries) {
      if (!e.name) continue
      const src = `${srcPath}/${e.name}`
      const dest = `${destPath}/${e.name}`
      if (e.isDirectory) {
        await handleFsCopyFolder(src, dest)
      } else {
        await copyFile(src, dest)
      }
    }
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleFsListAll(rootPath: string, maxFiles = 500) {
  const files: string[] = []
  await listAllFlat(rootPath, files, maxFiles)
  return { files }
}

export async function handleFsSearch(rootPath: string, query: string, maxResults = 50, caseSensitive = false) {
  const files: string[] = []
  await listAllFlat(rootPath, files, 2000)
  const q = caseSensitive ? query : query.toLowerCase()
  const results: Array<{ file: string; line: number; text: string }> = []
  const TEXT_EXT = new Set(['ts','tsx','js','jsx','py','go','c','cpp','md','txt','json','yaml','yml','sh','css','html'])
  outer: for (const f of files) {
    const ext = f.split('.').pop()?.toLowerCase() ?? ''
    if (!TEXT_EXT.has(ext)) continue
    let content: string
    try { content = await readTextFile(f) } catch { continue }
    const lines = content.split('\n')
    for (let i = 0; i < lines.length; i++) {
      const line = caseSensitive ? lines[i] : lines[i].toLowerCase()
      if (line.includes(q)) {
        results.push({ file: f, line: i + 1, text: lines[i].trim() })
        if (results.length >= maxResults) break outer
      }
    }
  }
  return { results }
}

export async function handleFsScanImports(rootPath: string) {
  const files: string[] = []
  await listAllFlat(rootPath, files, 500)
  const tsFiles = files.filter(f => /\.(ts|tsx|js|jsx)$/.test(f))
  const nodeMap = new Map<string, { id: string; label: string; filepath: string; type: string; is_main: boolean; x: number; y: number }>()
  const edges: Array<{ id: string; source: string; target: string }> = []
  let ei = 0
  for (const f of tsFiles) {
    const name = f.split('/').pop() ?? f
    const id = `n${nodeMap.size}`
    const isMain = /^(index|main|app)\.(ts|tsx|js|jsx)$/.test(name)
    nodeMap.set(f, {
      id, label: name, filepath: f,
      type: isMain ? 'entry' : 'function',
      is_main: isMain,
      x: (Math.random() - 0.5) * 800,
      y: (Math.random() - 0.5) * 600,
    })
  }
  for (const f of tsFiles) {
    let code: string
    try { code = await readTextFile(f) } catch { continue }
    const src = nodeMap.get(f)
    if (!src) continue
    const deps = extractImports(code, f)
    for (const dep of deps) {
      for (const [targetPath, tgt] of nodeMap) {
        if (targetPath.startsWith(dep) || dep.startsWith(targetPath.replace(/\.[^.]+$/, ''))) {
          edges.push({ id: `e${ei++}`, source: src.id, target: tgt.id })
          break
        }
      }
    }
  }
  return { nodes: [...nodeMap.values()], edges }
}

// ── Scripts getter ─────────────────────────────────────────────

export async function handleFsGetScripts(rootPath: string) {
  try {
    const pkgPath = `${rootPath}/package.json`
    if (!(await exists(pkgPath))) return { scripts: {} }
    const raw = await readTextFile(pkgPath)
    const pkg = JSON.parse(raw)
    return { scripts: pkg.scripts ?? {} }
  } catch {
    return { scripts: {} }
  }
}

// ── Code formatter (basic) ─────────────────────────────────────

export async function handleFsFormat(code: string, lang: string) {
  // No external formatter available on Android; return code as-is
  return { formatted: code, error: null }
}

// ── isomorphic-git FS adapter (used by git.ts) ────────────────

export const gitFsAdapter = {
  promises: {
    readFile: async (path: string, options?: any): Promise<Uint8Array | string> => {
      try {
        if (options?.encoding === 'utf8' || options?.encoding === 'utf-8') {
          return await readTextFile(path)
        }
        return await readFile(path)
      } catch (e: any) {
        const err: any = new Error(e?.message ?? 'ENOENT')
        err.code = 'ENOENT'
        throw err
      }
    },
    writeFile: async (path: string, data: Uint8Array | string, _options?: any): Promise<void> => {
      if (typeof data === 'string') {
        await writeTextFile(path, data)
      } else {
        await writeFile(path, data)
      }
    },
    unlink: async (path: string): Promise<void> => {
      await remove(path)
    },
    readdir: async (path: string): Promise<string[]> => {
      try {
        const entries = await readDir(path)
        return entries.map(e => e.name).filter(Boolean) as string[]
      } catch (e: any) {
        const err: any = new Error(e?.message ?? 'ENOENT')
        err.code = 'ENOENT'
        throw err
      }
    },
    mkdir: async (path: string, _options?: any): Promise<void> => {
      await mkdir(path, { recursive: true })
    },
    rmdir: async (path: string): Promise<void> => {
      await remove(path, { recursive: true })
    },
    stat: async (path: string): Promise<any> => {
      try {
        const s = await stat(path)
        return {
          type: s.isDirectory ? 'dir' : 'file',
          size: s.size ?? 0,
          mode: s.isDirectory ? 0o40755 : 0o100644,
          mtimeMs: s.mtime?.getTime() ?? Date.now(),
          ctimeMs: s.mtime?.getTime() ?? Date.now(),
          isFile: () => s.isFile,
          isDirectory: () => s.isDirectory,
          isSymbolicLink: () => s.isSymlink,
        }
      } catch (e: any) {
        const err: any = new Error(e?.message ?? 'ENOENT')
        err.code = 'ENOENT'
        throw err
      }
    },
    lstat: async (path: string): Promise<any> => {
      try {
        const s = await stat(path)
        return {
          type: s.isDirectory ? 'dir' : 'file',
          size: s.size ?? 0,
          mode: s.isDirectory ? 0o40755 : 0o100644,
          mtimeMs: s.mtime?.getTime() ?? Date.now(),
          ctimeMs: s.mtime?.getTime() ?? Date.now(),
          isFile: () => s.isFile,
          isDirectory: () => s.isDirectory,
          isSymbolicLink: () => s.isSymlink,
        }
      } catch (e: any) {
        const err: any = new Error(e?.message ?? 'ENOENT')
        err.code = 'ENOENT'
        throw err
      }
    },
    rename: async (oldPath: string, newPath: string): Promise<void> => {
      await rename(oldPath, newPath)
    },
    symlink: async (): Promise<void> => {
      throw Object.assign(new Error('symlinks not supported'), { code: 'ENOSYS' })
    },
    readlink: async (): Promise<string> => {
      throw Object.assign(new Error('symlinks not supported'), { code: 'ENOSYS' })
    },
    chmod: async (): Promise<void> => { /* no-op on Android */ },
  },
}

File: engine/fs.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
	"unicode/utf8"
)

// ── Filesystem helpers ─────────────────────────────────────────

// cleanPath resolves any ../ segments and symlinks so callers
// can't escape the workspace with a crafted path like
// /project/../../../etc/passwd.
func cleanPath(root, target string) (string, error) {
	abs, err := filepath.Abs(target)
	if err != nil {
		return "", err
	}
	if root != "" {
		rootAbs, _ := filepath.Abs(root)
		if !strings.HasPrefix(abs+string(filepath.Separator), rootAbs+string(filepath.Separator)) {
			return "", fmt.Errorf("path escapes workspace")
		}
	}
	return abs, nil
}

// ── Filesystem handlers ────────────────────────────────────────

type FsNode struct {
	Name     string   `json:"name"`
	Path     string   `json:"path"`
	Kind     string   `json:"type"`
	Ext      string   `json:"ext,omitempty"`
	Children []FsNode `json:"children,omitempty"`
}

func buildTree(p string, depth, maxDepth int) FsNode {
	name := filepath.Base(p)
	ext := strings.ToLower(filepath.Ext(p))
	info, err := os.Stat(p)
	if err != nil || !info.IsDir() {
		return FsNode{Name: name, Path: p, Kind: "file", Ext: ext}
	}
	node := FsNode{Name: name, Path: p, Kind: "dir"}
	if depth < maxDepth {
		entries, err := os.ReadDir(p)
		if err == nil {
			var dirs, files []FsNode
			for _, e := range entries {
				if shouldIgnore(e.Name()) {
					continue
				}
				child := buildTree(filepath.Join(p, e.Name()), depth+1, maxDepth)
				if e.IsDir() {
					dirs = append(dirs, child)
				} else {
					files = append(files, child)
				}
			}
			node.Children = append(dirs, files...)
		}
	}
	if node.Children == nil {
		node.Children = []FsNode{}
	}
	return node
}

func handleFsTree(w http.ResponseWriter, r *http.Request) {
	var req struct {
		RootPath string `json:"rootPath"`
		MaxDepth *int   `json:"maxDepth"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	root, err := cleanPath("", req.RootPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	req.RootPath = root
	maxDepth := 6
	if req.MaxDepth != nil {
		maxDepth = *req.MaxDepth
	}
	tree := buildTree(req.RootPath, 0, maxDepth)
	jsonResp(w, map[string]any{"success": true, "tree": tree})
}

func handleFsRead(w http.ResponseWriter, r *http.Request) {
	var req struct {
		FilePath string `json:"filePath"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	fp, err := cleanPath("", req.FilePath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	req.FilePath = fp
	info, err := os.Stat(req.FilePath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	if info.Size() > 50*1024*1024 {
		jsonResp(w, map[string]any{"success": false, "error": "file too large (>50 MB)", "tooLarge": true})
		return
	}
	content, err := os.ReadFile(req.FilePath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	if !utf8.Valid(content) {
		jsonResp(w, map[string]any{"success": false, "error": "file is binary"})
		return
	}
	jsonResp(w, map[string]any{"success": true, "content": string(content)})
}

func handleFsWrite(w http.ResponseWriter, r *http.Request) {
	var req struct {
		FilePath string `json:"filePath"`
		Content  string `json:"content"`
	}
	if !decodeJSON(w, r, &req) {
		return
	}
	fp, err := cleanPath("", req.FilePath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	req.FilePath = fp
	if dir := filepath.Dir(req.FilePath); dir != "" {
		os.MkdirAll(dir, 0755)
	}
	if err := os.WriteFile(req.FilePath, []byte(req.Content), 0644); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleFsCreateFile(w http.ResponseWriter, r *http.Request) {
	var req struct {
		FilePath string `json:"filePath"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	fp, err := cleanPath("", req.FilePath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	req.FilePath = fp
	if _, err := os.Stat(req.FilePath); err == nil {
		jsonResp(w, map[string]any{"success": false, "error": "File already exists"})
		return
	}
	if dir := filepath.Dir(req.FilePath); dir != "" {
		os.MkdirAll(dir, 0755)
	}
	if err := os.WriteFile(req.FilePath, []byte{}, 0644); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleFsCreateDir(w http.ResponseWriter, r *http.Request) {
	var req struct {
		FolderPath string `json:"folderPath"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	fp, err := cleanPath("", req.FolderPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	req.FolderPath = fp
	if err := os.MkdirAll(req.FolderPath, 0755); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleFsDelete(w http.ResponseWriter, r *http.Request) {
	var req struct {
		ItemPath string `json:"itemPath"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	ip, err := cleanPath("", req.ItemPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	req.ItemPath = ip
	info, err := os.Stat(req.ItemPath)
	var rerr error
	if err == nil && info.IsDir() {
		rerr = os.RemoveAll(req.ItemPath)
	} else {
		rerr = os.Remove(req.ItemPath)
	}
	if rerr != nil {
		jsonResp(w, map[string]any{"success": false, "error": rerr.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleFsRename(w http.ResponseWriter, r *http.Request) {
	var req struct {
		OldPath string `json:"oldPath"`
		NewPath string `json:"newPath"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	op, err := cleanPath("", req.OldPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	np, err := cleanPath("", req.NewPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	req.OldPath, req.NewPath = op, np
	if err := os.Rename(req.OldPath, req.NewPath); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleFsCopyFile(w http.ResponseWriter, r *http.Request) {
	var req struct {
		SrcPath  string `json:"srcPath"`
		DestPath string `json:"destPath"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	sp, err := cleanPath("", req.SrcPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	dp, err := cleanPath("", req.DestPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	req.SrcPath, req.DestPath = sp, dp
	data, err := os.ReadFile(req.SrcPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	if err := os.WriteFile(req.DestPath, data, 0644); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleFsCopyFolder(w http.ResponseWriter, r *http.Request) {
	var req struct {
		SrcPath  string `json:"srcPath"`
		DestPath string `json:"destPath"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	sp, err := cleanPath("", req.SrcPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	dp, err := cleanPath("", req.DestPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	req.SrcPath, req.DestPath = sp, dp
	var copyDir func(src, dst string) error
	copyDir = func(src, dst string) error {
		if err := os.MkdirAll(dst, 0755); err != nil {
			return err
		}
		entries, err := os.ReadDir(src)
		if err != nil {
			return err
		}
		for _, e := range entries {
			sp := filepath.Join(src, e.Name())
			dp := filepath.Join(dst, e.Name())
			if e.IsDir() {
				if err := copyDir(sp, dp); err != nil {
					return err
				}
			} else {
				data, err := os.ReadFile(sp)
				if err != nil {
					return err
				}
				if err := os.WriteFile(dp, data, 0644); err != nil {
					return err
				}
			}
		}
		return nil
	}
	if err := copyDir(req.SrcPath, req.DestPath); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleFsListAll(w http.ResponseWriter, r *http.Request) {
	var req struct {
		RootPath string `json:"rootPath"`
		MaxFiles *int   `json:"maxFiles"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	limit := 5000
	if req.MaxFiles != nil {
		limit = *req.MaxFiles
	}
	results := []map[string]any{}
	var walk func(dir, rel string)
	walk = func(dir, rel string) {
		if len(results) >= limit {
			return
		}
		entries, err := os.ReadDir(dir)
		if err != nil {
			return
		}
		for _, e := range entries {
			n := e.Name()
			if shouldIgnore(n) {
				continue
			}
			full := filepath.Join(dir, n)
			relPath := n
			if rel != "" {
				relPath = rel + "/" + n
			}
			if e.IsDir() {
				walk(full, relPath)
			} else {
				results = append(results, map[string]any{"path": full, "rel": relPath, "name": n})
			}
		}
	}
	walk(req.RootPath, "")
	jsonResp(w, results)
}

func handleFsSearch(w http.ResponseWriter, r *http.Request) {
	var req struct {
		RootPath      string `json:"rootPath"`
		Query         string `json:"query"`
		MaxResults    *int   `json:"maxResults"`
		CaseSensitive bool   `json:"caseSensitive"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	sr, err := cleanPath("", req.RootPath)
	if err != nil {
		jsonResp(w, map[string]any{"error": err.Error()})
		return
	}
	req.RootPath = sr
	if len(req.Query) < 2 {
		jsonResp(w, []any{})
		return
	}
	limit := 300
	if req.MaxResults != nil {
		limit = *req.MaxResults
	}
	lower := strings.ToLower(req.Query)
	textExts := map[string]bool{
		".js": true, ".ts": true, ".tsx": true, ".jsx": true,
		".py": true, ".go": true, ".c": true, ".cpp": true,
		".h": true, ".md": true, ".json": true, ".css": true,
		".html": true, ".txt": true, ".yaml": true, ".yml": true,
		".toml": true, ".rs": true, ".sh": true,
	}
	ctx, cancel := context.WithTimeout(r.Context(), 8*time.Second)
	defer cancel()

	results := []map[string]any{}
	var walk func(dir, rel string)
	walk = func(dir, rel string) {
		if len(results) >= limit {
			return
		}
		select {
		case <-ctx.Done():
			return
		default:
		}
		entries, err := os.ReadDir(dir)
		if err != nil {
			return
		}
		for _, e := range entries {
			n := e.Name()
			if shouldIgnore(n) {
				continue
			}
			full := filepath.Join(dir, n)
			relPath := n
			if rel != "" {
				relPath = rel + "/" + n
			}
			if e.IsDir() {
				walk(full, relPath)
			} else {
				ext := strings.ToLower(filepath.Ext(n))
				if !textExts[ext] {
					continue
				}
				data, err := os.ReadFile(full)
				if err != nil {
					continue
				}
				for i, line := range strings.Split(string(data), "\n") {
					var col int
					if req.CaseSensitive {
						col = strings.Index(line, req.Query)
					} else {
						col = strings.Index(strings.ToLower(line), lower)
					}
					if col >= 0 {
						text := strings.TrimSpace(line)
						if len(text) > 200 {
							text = text[:200]
						}
						results = append(results, map[string]any{
							"file": relPath, "fullPath": full,
							"line": i + 1, "text": text, "col": col,
						})
						if len(results) >= limit {
							return
						}
					}
				}
			}
		}
	}
	walk(req.RootPath, "")
	jsonResp(w, results)
}

func handleFsGetScripts(w http.ResponseWriter, r *http.Request) {
	var req struct {
		RootPath string `json:"rootPath"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	data, err := os.ReadFile(filepath.Join(req.RootPath, "package.json"))
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "scripts": []any{}, "type": "none"})
		return
	}
	var pkg map[string]any
	if err := json.Unmarshal(data, &pkg); err != nil {
		jsonResp(w, map[string]any{"success": false, "scripts": []any{}, "type": "none"})
		return
	}
	scripts, ok := pkg["scripts"].(map[string]any)
	if !ok {
		jsonResp(w, map[string]any{"success": false, "scripts": []any{}, "type": "none"})
		return
	}
	list := make([]map[string]any, 0, len(scripts))
	for k, v := range scripts {
		list = append(list, map[string]any{"name": k, "cmd": v, "source": "npm"})
	}
	name, _ := pkg["name"].(string)
	jsonResp(w, map[string]any{"success": true, "scripts": list, "type": "npm", "name": name})
}

func handleFsFormat(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Code string `json:"code"`
		Lang string `json:"lang"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	extMap := map[string]string{
		"go": "go", "py": "py", "js": "js", "jsx": "js",
		"ts": "ts", "tsx": "ts", "css": "css", "json": "json", "html": "html",
	}
	ext, ok := extMap[req.Lang]
	if !ok {
		jsonResp(w, map[string]any{"success": false, "error": "No formatter"})
		return
	}
	tmp, err := os.CreateTemp("", "forbiden_fmt_*."+ext)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	tmpPath := tmp.Name()
	tmp.WriteString(req.Code)
	tmp.Close()
	defer os.Remove(tmpPath)

	var cmdStr string
	switch req.Lang {
	case "go":
		cmdStr = "gofmt -w " + tmpPath
	case "py":
		cmdStr = "black " + tmpPath + " 2>&1 || autopep8 --in-place " + tmpPath
	default:
		cmdStr = "npx --yes prettier --write " + tmpPath
	}
	cmd := exec.Command("sh", "-c", cmdStr)
	cmd.Env = append(os.Environ(), "PATH="+extendedPath())
	if err := cmd.Run(); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": "Format failed"})
		return
	}
	result, err := os.ReadFile(tmpPath)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true, "code": string(result)})
}

func handleFsScanImports(w http.ResponseWriter, r *http.Request) {
	var req struct {
		RootPath string `json:"rootPath"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	codeExts := map[string]bool{
		"js": true, "jsx": true, "ts": true, "tsx": true, "py": true,
		"go": true, "c": true, "cpp": true, "h": true, "hpp": true,
		"rs": true, "rb": true, "java": true, "kt": true, "swift": true, "cs": true,
	}
	var files []string
	filepath.WalkDir(req.RootPath, func(path string, d os.DirEntry, err error) error {
		if err != nil || d == nil {
			return nil
		}
		n := d.Name()
		if d.IsDir() {
			if shouldIgnore(n) {
				return filepath.SkipDir
			}
			return nil
		}
		ext := strings.ToLower(strings.TrimPrefix(filepath.Ext(n), "."))
		if codeExts[ext] {
			files = append(files, path)
		}
		return nil
	})
	n := len(files)
	nodes := make([]map[string]any, n)
	for i, f := range files {
		rel, _ := filepath.Rel(req.RootPath, f)
		ext := strings.ToLower(strings.TrimPrefix(filepath.Ext(f), "."))
		angle := float64(i) / float64(max(n, 1)) * 2 * math.Pi
		radius := 200.0
		nodes[i] = map[string]any{
			"id": fmt.Sprintf("fi%d", i), "label": rel, "path": f,
			"ext": ext, "type": "function", "themeIdx": i % 16,
			"x": radius * math.Cos(angle), "y": radius * math.Sin(angle),
		}
	}
	jsonResp(w, map[string]any{
		"success": true, "nodes": nodes, "edges": []any{},
		"rootPath": req.RootPath, "fileCount": n,
	})
}
